File: array/longest_non_repeat/longest_non_repeat.rb
def longest_non_repeat(string)
  start = 0
  maxlen = 0
  used_char = {}
  string.each_char.with_index do |char, i|
    symb = char.to_sym
    if used_char.key?(symb) && (start <= used_char[symb])
      start = used_char[symb] + 1
    else
      maxlen = [maxlen, i + 1 - start].max
    end
    used_char[symb] = i
  end
  maxlen
end

a = 'abcabcdefbb'
p("input: #{a}")
p("result: #{longest_non_repeat(a)}")
p('output should be 6')

File: array/longest_non_repeat/longest_non_repeat.js
function longestNonRepeat(s) {
  let start = 0
  let maxlen = 0
  const usedChar = new Map()
  s.split('').forEach((char, i) => {
    if (usedChar.has(char) && start <= usedChar.get(char)) {
      start = usedChar.get(char) + 1
    } else {
      maxlen = Math.max(maxlen, i + 1 - start)
    }
    usedChar.set(char, i)
  })
  return maxlen
}

const a = 'abcabcdefbb'
console.log('input: ', a)
console.log('result: ', longestNonRepeat(a))
console.log('output should be 6')
